File: apps/api/app/schemas/budget.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.schemas.common import PaginationMeta


class BudgetLineItemCreate(BaseModel):
    cost_code: str
    description: str
    original_amount: int  # Cents
    notes: Optional[str] = None


class BudgetLineItemUpdate(BaseModel):
    description: Optional[str] = None
    original_amount: Optional[int] = None  # Cents
    notes: Optional[str] = None


class BudgetLineItemResponse(BaseModel):
    id: UUID
    project_id: UUID
    cost_code: str
    description: Optional[str] = None
    original_amount: int  # Cents
    approved_changes: int  # Cents
    revised_amount: int  # Cents (original + approved_changes)
    billed_to_date: int  # Cents
    remaining: int  # Cents (revised - billed)
    percent_complete: float
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class BudgetSummaryResponse(BaseModel):
    original_contract: int  # Cents
    approved_changes: int  # Cents
    revised_contract: int  # Cents
    billed_to_date: int  # Cents
    remaining: int  # Cents
    percent_complete: float
    line_items: list[BudgetLineItemResponse]
    change_orders_pending: int
    change_orders_pending_amount: int  # Cents


class BudgetImportItem(BaseModel):
    cost_code: str
    description: str
    amount: int  # Cents

File: apps/web/src/types/budget.ts
export interface BudgetLineItem {
  id: string;
  project_id: string;
  cost_code: string;
  description: string | null;
  original_amount: number; // cents
  approved_changes: number; // cents
  revised_amount: number; // cents
  billed_to_date: number; // cents
  remaining: number; // cents
  percent_complete: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface BudgetSummary {
  original_contract: number; // cents
  approved_changes: number; // cents
  revised_contract: number; // cents
  billed_to_date: number; // cents
  remaining: number; // cents
  percent_complete: number;
  line_items: BudgetLineItem[];
  change_orders_pending: number;
  change_orders_pending_amount: number; // cents
}

export interface BudgetLineItemCreateInput {
  cost_code: string;
  description: string;
  original_amount: number; // cents
  notes?: string;
}

export interface BudgetLineItemUpdateInput {
  description?: string;
  original_amount?: number; // cents
  notes?: string;
}
